src/lib.rs

```diff
@@ -6,231 +6,108 @@ extern crate sha3;
 pub mod instruction;
 pub mod multiproof;
 pub mod node;
+pub mod tree;
 pub mod utils;
 
 use instruction::*;
 use multiproof::*;
 use node::*;
+use tree::{KeyValueStore, Tree};
 use utils::*;
 
-// Rebuilds the tree based on the multiproof components
-pub fn rebuild(proof: &Multiproof) -> Result<Node, String> {
-    use Instruction::*;
-    use Node::*;
+impl<S: KeyValueStore, T: Tree<S> + rlp::Decodable> ProofToTree<S, T> for Multiproof {
+    fn rebuild(&self) -> Result<T, String> {
+        use Instruction::*;
 
-    let mut hiter = proof.hashes.iter();
-    let iiter = proof.instructions.iter();
-    let mut kviter = proof.keyvals.iter().map(|encoded| {
-        // Deserialize the keys as they are read
-        rlp::decode::<Node>(encoded).unwrap()
-    });
+        let mut hiter = self.hashes.iter();
+        let iiter = self.instructions.iter();
+        let mut kviter = self.keyvals.iter().map(|encoded| {
+            // Deserialize the keys as they are read
+            rlp::decode::<T>(encoded).unwrap()
+        });
 
-    let mut stack = Vec::<Node>::new();
+        let mut stack = Vec::<T>::new();
 
-    for instr in iiter {
-        match instr {
-            HASHER => {
-                if let Some(h) = hiter.next() {
-                    stack.push(Hash(h.to_vec()));
-                } else {
-                    return Err(format!("Proof requires one more hash in HASHER"));
+        for instr in iiter {
+            match instr {
+                HASHER => {
+                    if let Some(h) = hiter.next() {
+                        stack.push(T::from_hash(h.to_vec()));
+                    } else {
+                        return Err(format!("Proof requires one more hash in HASHER"));
+                    }
                 }
-            }
-            LEAF(keylength) => {
-                if let Some(Leaf(key, value)) = kviter.next() {
-                    // If the value is empty, we have a NULL key and
-                    // therefore an EmptySlot should be returned.
-                    if value.len() != 0 {
-                        stack.push(Leaf(
-                            NibbleKey::from(key[key.len() - *keylength..].to_vec()),
-                            value.to_vec(),
-                        ));
+                LEAF(keylength) => {
```

Comment:
Ah `LEAF`'s operand has been deprecated as in it's not used, but it still has an operand, I misread your comment initially. This makes it easy to remove the operand later thanks :+1: